drivers/mfd/mfd_ite_it8801.c

```diff
@@ -0,0 +1,143 @@
+/*
+ * Copyright (c) 2024 ITE Corporation. All Rights Reserved.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT ite_it8801_mfd
+
+#include <zephyr/drivers/gpio.h>
+#include <zephyr/drivers/i2c.h>
+#include <zephyr/drivers/mfd/mfd_ite_it8801.h>
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(mfd_ite_it8801, CONFIG_MFD_LOG_LEVEL);
+
+struct mfd_it8801_config {
+	const struct i2c_dt_spec i2c_dev;
+	/* Alert GPIO pin */
+	const struct gpio_dt_spec irq_gpios;
+	/* MFD sub device */
+	const struct device **sub_mfd_dev;
+	/* MFD sub device count */
+	uint8_t sub_mfd_node_num;
+};
+
+struct mfd_it8801_data {
+	struct k_work gpio_isr_worker;
+	/* Alert pin callback */
+	struct gpio_callback gpio_cb;
+	const struct device *mfd_dev;
+};
+
+static int it8801_check_vendor_id(const struct device *dev)
+{
+	const struct mfd_it8801_config *config = dev->config;
+	int i, ret;
+	uint8_t val;
+
+	/*  Verify vendor ID registers(16-bits). */
+	for (i = 0; i < ARRAY_SIZE(it8801_id_verify); i++) {
+		ret = i2c_reg_read_byte_dt(&config->i2c_dev, it8801_id_verify[i].reg, &val);
+
+		if (ret != 0) {
+			LOG_ERR("Failed to read vendoer ID (ret %d)", ret);
+			return ret;
+		}
+
+		if (val != it8801_id_verify[i].chip_id) {
+			LOG_ERR("The IT8801 vendor ID is wrong");
+			return -ENODEV;
+		}
+	}
+
+	return ret;
+}
+
+void it8801_gpio_callback(const struct device *dev, struct gpio_callback *cb, uint32_t pins)
+{
+	ARG_UNUSED(pins);
+	struct mfd_it8801_data *data = CONTAINER_OF(cb, struct mfd_it8801_data, gpio_cb);
+
+	k_work_submit(&data->gpio_isr_worker);
+}
+
+static void it8801_gpio_alert_worker(struct k_work *work)
+{
+	struct mfd_it8801_data *data = CONTAINER_OF(work, struct mfd_it8801_data, gpio_isr_worker);
+	const struct mfd_it8801_config *config = data->mfd_dev->config;
+	int i;
+
+	for (i = 0; i < config->sub_mfd_node_num; i++) {
+		/*
+		 * When the registered GPIO is triggered, the parent MFD driver
+		 * will scan all the subdevices, and call the subdevice handler
+		 * to determine which device was triggered.
+		 */
+#ifdef CONFIG_INPUT_ITE_IT8801_KBD
+		it8801_input_alert_handler(config->sub_mfd_dev[i]);
+#endif
+#ifdef CONFIG_GPIO_ITE_IT8801
+		it8801_gpio_alert_handler(config->sub_mfd_dev[i]);
+#endif
```

Comment:
Another option here.  The children devices already have a reference to the parent `mfd`.  The MFD driver could add an API to register an interrupt callback.

```
mfd_it8801_register_interrupt_callback(const struct device *mfd,
    const struct device *child,
    callback_function);
```
Then you could also drop `sub_mfd_dev` array, and you could drop the `if (dev != data->gpio_dev)` checks in the child drivers.

Reply:
Thanks for the suggestion! It really helped simplify the code.